EQL: Add integration tests harness to test EQL feature parity with original implementation

aleksmaus opened the pull request:
The tests use the original test queries from
https://github.com/endgameinc/eql/blob/master/eql/etc/test_queries.toml
for EQL implementation correctness validation.
The file test_queries_unsupported.toml serves as a "blacklist" for the
queries that we do not support. Currently all of the queries are
blacklisted. Over the time the expectation is to eventually have an
empty "blacklist" when all of the queries are fully supported.

The tests use the original test vector from
https://raw.githubusercontent.com/endgameinc/eql/master/eql/etc/test_data.json
that was translated to ECS format that matches the latest mapping being
used for Endgame platform event streaming and is loaded from endgame.dat
file. The endgame.json file contains the matching index
mappings/setting.

Only one EQL spec response is stubbed for now to match the expected
output for that spec query. This part would need some tweaking after EQL is
fully wired.

The input .toml file is parsed by hand for now, which is sufficient for
our purposes and avoids introducing another dependency just for this
particular test case.

Related to https://github.com/elastic/elasticsearch/issues/49581

@imotov could you take a look first before I promote this from draft PR?
Thanks!

elasticmachine commented:
Pinging @elastic/es-search (:Search/EQL)

aleksmaus commented:
@elasticmachine run elasticsearch-ci/packaging-sample-matrix-unix

rw-access commented on x-pack/plugin/eql/src/test/resources/endgame.dat:
```diff
@@ -0,0 +1,102 @@
+{"user":{"group":{}},"host":{"os":{"platform":"windows","name":"Windows"},"ip":"127.0.0.1","hostname":"localhost","name":"localhost"},"event":{"module":"endgame","dataset":"esensor","action":"already_running","category":"process","kind":"event"},"labels":{"account_id":"f6c123dc-b6d9-4849-937b-08c444ce7d96","endpoint_id":"6a19472c-f0ab-4f3a-b83d-dc3698a5ff66"},"agent":{"type":"endgame","id":"6a194 ...
```
two q's
- why the `.dat` extension?
- I don't think this needs to be in ECS, other than the event.category field. Otherwise all of the queries need to be rewritten

aleksmaus replied:
1. it's not quite a valid json, it's a collection of jsons. open for suggestions on the file extension.
2. wanted to use the exact same ECS data format that is currently established for events streaming from SMP, so we could potentially flash out the details.

aleksmaus replied:
Updated to .ndjson

rw-access commented on x-pack/plugin/eql/src/test/resources/test_queries_unsupported.toml:
```diff
@@ -0,0 +1,1300 @@
+# official EQL implementation
+
+[[queries]]
+query = 'process where serial_event_id < 4'
```
this one is supported

aleksmaus replied:
yeah, but not hooked up to the query engine yet. Once it is, we can validate the results and update the .toml files.

aleksmaus replied:
expecting the test_queries_unsupported.toml file to become empty and eventually dropped when we reach the feature parity

rw-access replied:
oh, I didn't realize that. in that case, this file is missing the first query from test_queries.toml. i just looked at the top and saw the discrepancy and reached the wrong conclusion
```toml
[[queries]]
query = 'process where serial_event_id = 1'
expected_event_ids  = [1]
```

aleksmaus replied:
Added the comment into the file per @rw-access recommendation.

imotov commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java:
```diff
@@ -0,0 +1,95 @@
+
+    @BeforeClass
+    public static void checkForSnapshot() {
+        assumeTrue("Only works on snapshot builds for now", Build.CURRENT.isSnapshot());
```
I don't think we need this anymore.

aleksmaus commented:
@imotov,  I talked to @costin before and got few suggestions. Basically, wanted to keep the original test spec from EQL unchanged, so we can easily follow any original EQL implementation changes by copying the toml file over into our test harness. Also tried to avoid adding any toml library dependency, since the format of the file is pretty simple to parse as is without the need for a full toml parser implementation yet, especially for testing.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java:
```diff
@@ -103,9 +102,24 @@
     public List<Setting<?>> getSettings() {
```
Either we use `enabled` or this method. On one hand `enabled` checks the property presence statically but this method tries to do it dynamic when a class is instantiated.
There's there are multiple pieces of `isSnapshot` in this class that need simplification.

aleksmaus replied:
tried to follow the autoscaling plugin pattern, that I was told should be an example of handling the feature flag.

costin commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/TransportEqlSearchAction.java:
```diff
@@ -41,7 +40,7 @@
     @Inject
     public TransportEqlSearchAction(Settings settings, ClusterService clusterService, TransportService transportService,
-            ThreadPool threadPool, ActionFilters actionFilters, PlanExecutor planExecutor) {
+                                    ThreadPool threadPool, ActionFilters actionFilters, PlanExecutor planExecutor) {
```
Same as above, formatting on non-modified lines is noise.

aleksmaus replied:
this is the formatting misalignment that I missed in my previous PR, now trying to fix it

costin commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+
+    private static String[] readArray(String line) throws Exception {
+        line = readValueLine(line);
+        if (!line.startsWith("[") && !line.endsWith("]")) {
```
nit: the preference in general is not to negate but instead use `== false` for clarity.

aleksmaus replied:
will update

aleksmaus replied:
this is no longer relevant, replacing with toml library

costin commented on x-pack/plugin/sql/build.gradle:
```diff
@@ -26,6 +26,10 @@ configurations {
 archivesBaseName = 'x-pack-sql'
 
+test {
+  testLogging.showStandardStreams = true
```
Why?

aleksmaus replied:
will remove

aleksmaus commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java:
```diff
@@ -69,31 +72,27 @@
-    public Collection<Object> createComponents(Client client, ClusterService clusterService, ThreadPool threadPool,
-            ResourceWatcherService resourceWatcherService, ScriptService scriptService, NamedXContentRegistry xContentRegistry,
-            Environment environment, NodeEnvironment nodeEnvironment, NamedWriteableRegistry namedWriteableRegistry) {
+    public Collection<Object> createComponents(Client client, ClusterService clusterService,
```
Just making it easier to read, also because it just was just brought to my attention that some of the code that I had some formatting issues.

aleksmaus commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java:
```diff
@@ -103,9 +102,24 @@
+    }
+
+    @Override
+    public List<ActionHandler<? extends ActionRequest, ? extends ActionResponse>> getActions() {
```
consistent with other plugins. can undo

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java:
```diff
@@ -0,0 +1,95 @@
+            while ((line = reader.readLine()) != null) {
+                bulkBuilder.add(new IndexRequest(testIndexName).source(line.trim(), XContentType.JSON));
+            }
+            BulkResponse response = bulkBuilder.setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).get();
```
Import static for `IMMEDIATE`?

aleksmaus replied:
Seems like this is consistent with 100+ other places 
![Screen Shot 2020-02-18 at 2 00 58 PM](https://user-images.githubusercontent.com/872351/74768741-1b303d00-5257-11ea-91cc-18533d3ba4ae.png)

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java:
```diff
@@ -0,0 +1,95 @@
+            boolean supported = true;
+            // Check if spec is supported, simple iteration, cause the list is short.
+            for (EqlSpec unSpec : unsupportedSpecs) {
+                if (spec.query.equals(unSpec.query)) {
```
Any reason why not making `query` private inside `EqlSpec` and provide a `query()` accessor method?

aleksmaus replied:
Less typing for the test code. Will convert to accessors.

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+import java.util.Map;
+
+public class EqlSpecLoader {
+    public static List<EqlSpec> load(String path, boolean supported) throws Exception {
```
Will this method be used in the future? (atm it isn't)

aleksmaus replied:
EqlSpecLoader.load is used here 
https://github.com/elastic/elasticsearch/pull/52248/files#diff-663165787e62c937a0aa8ac3d8511d6dR63

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+        }
+    }
+
+    private static int[] readExpectedEventIds(String line, BufferedReader reader) throws Exception {
```
Imo, it would help to see a method comment containing sample line(s) these `read*` methods are expected to handle.

aleksmaus replied:
No longer relevant, replaced with toml parser

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+            if (s.endsWith("\"") || s.endsWith("'")) {
+                s = s.substring(0, s.length() - 1);
+            }
+            arr[i] = s.trim();
```
Why are all these `read*` methods trimming the values they are reading? For example, `readTags` gets its values from `readArray` which is already trimming the values returned as array...
Are all these `trim()` calls really needed or they are added just "to be safe"?

aleksmaus replied:
in this particular case the trimming is done for the case where the value in the brackets prefixed/suffixed with spaces, for example: " foo bar "

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+
+    private static String readString(String line, BufferedReader reader) throws Exception {
+        line = readValueLine(line);
+        String delim = "";
```
I don't think this initialization is necessary. A simple `String delim;` should be enough.

aleksmaus replied:
IDE was complaining on unitialized var delim otherwise

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+        List<EqlSpec> testSpecs = new ArrayList<>();
+
+        EqlSpec spec = null;
+        try (BufferedReader reader = new BufferedReader(
```
Single `try()` line.

aleksmaus replied:
No longer relevant, replaced with toml parser.

nik9000 commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java:
```diff
@@ -0,0 +1,95 @@
+                }
+            }
+
+            if (supported) {
```
I think this test should use `@ParametersFactory` so each example becomes a test that we can address individually.

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+            }
+        }
+
+        if (StringUtil.isNullOrEmpty(delim)) {
```
There is a class `org.elasticsearch.common.Strings` that has the same method and I think we prefer that one.

aleksmaus replied:
No longer relevant, replaced with toml parser.

astefan commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlSpecLoader.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private static void validateAndAddSpec(List<EqlSpec> specs, EqlSpec spec, boolean supported) throws Exception {
+        if (StringUtil.isNullOrEmpty(spec.query)) {
```
Same here about `Strings` class.

aleksmaus replied:
Updated

astefan commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/action/EqlSearchResponse.java:
```diff
@@ -71,6 +70,7 @@
  *         EqlSearchResponse.Hits hits = new EqlSearchResponse.Hits(null, null, counts, totals);
  *         EqlSearchResponse response = new EqlSearchResponse(hits, 5, false);
  */
+
```
Can you remove this, please?

aleksmaus replied:
Sure. It was initially useful to have while we were iterating on the response shape.

aleksmaus commented on x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/plugin/EqlPlugin.java:
```diff
@@ -103,9 +102,24 @@
+            );
+        }
+        return List.of(
+            new ActionHandler<>(XPackUsageFeatureAction.EQL, EqlUsageTransportAction.class),
```
I would not sweat about this particular code, it will be cleaner once we remove the feature flag.

aleksmaus commented:
@nik9000 replacing the manual toml parsing with io.ous:jtoml:2.0.0.
The other couple of libraries that I looked at depend on antlr4  4.7.1 which breaks compatibility with existing 4.5.3.

aleksmaus commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/EqlTestUtils.java:
```diff
@@ -16,19 +18,33 @@
+            randomAlphaOfLength(16));
+    }
+
+    public static Tuple<String, String> pathAndName(String string) {
```
No longer used indeed, will remove

aleksmaus commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/AbstractEqlIntegTestCase.java:
```diff
@@ -0,0 +1,41 @@
+    @Override
+    protected Settings nodeSettings(int nodeOrdinal) {
+        Settings.Builder settings = Settings.builder().put(super.nodeSettings(nodeOrdinal));
+        settings.put(XPackSettings.SECURITY_ENABLED.getKey(), false);
```
Probably. Was following AbstractSqlIntegTestCase as example in this case.

aleksmaus commented on x-pack/plugin/eql/src/test/java/org/elasticsearch/xpack/eql/action/EqlActionIT.java:
```diff
@@ -0,0 +1,95 @@
+        InputStream is = EqlActionIT.class.getResourceAsStream("/endgame.dat");
+        BulkRequestBuilder bulkBuilder = client().prepareBulk();
+        try (BufferedReader reader = new BufferedReader(
+            new InputStreamReader(is, StandardCharsets.UTF_8))) {
```
No longer relevant, replaced with the toml parser library

rw-access commented on x-pack/plugin/eql/src/test/resources/endgame.json:
```diff
@@ -0,0 +1,2531 @@
+{
```
nit: a more descriptive filename. i.e. endgame_mappings.json

rw-access commented on x-pack/plugin/eql/src/test/resources/endgame.ndjson:
```diff
@@ -0,0 +1,102 @@
+{"user":{"group":{}},"host":{"os":{"platform":"windows","name":"Windows"},"ip":"127.0.0.1","hostname":"localhost","name":"localhost"},"event":{"module":"endgame","dataset":"esensor","action":"already_running","category":"process","kind":"event"},"labels":{"account_id":"f6c123dc-b6d9-4849-937b-08c444ce7d96","endpoint_id":"6a19472c-f0ab-4f3a-b83d-dc3698a5ff66"},"agent":{"type":"endgame","id":"6a194 ...
```
the mappings are consistent with the data, but the queries are not. EQL is independent of ECS, so I think it's okay to leave the data in the same schema that it was in

aleksmaus replied:
using this format mostly in order to validate and work through any possible issues with covering our particular use case

rw-access replied:
that means we have to manually rewrite all of the queries though, which is likely more error prone.

aleksmaus replied:
Discussed with @rw-access on Slack. 
Originally I was under impression that we will keep the queries/rules the same and use some internal mapping to adopt them for ECS.  I looks like in the first iteration this will not be the case and the queries/rules will be rewritten for ECS.
I will update this PR to use the same flat json data from the original EQL tests and remove the ECS mapping template creation for now.

aleksmaus commented:
Updated PR with the latest recommendations from @rw-access, also including @nik9000 suggestion to use @ParametersFactory for the test harness.

aleksmaus commented:
Added a small adjustment after merging the latest master changes that included @imotov HL client test against the currently hardcoded EQL response. I changed the shape of the response for the first of the EQL validation harness test queries expected response, while Igor was working on HL client.

aleksmaus commented:
Had to merge master again and resolve conflicts. Will merge once it clears CI build.
